packages/hashtower.circom/circuits/hashtower-hash-chain.circom

```diff
@@ -0,0 +1,254 @@
+    // So out[] could only be len 1s followed by N - len 0s .
+    var oneCount = 0;
+    for (var i = 0; i < N; i++) {
+        out[i] <-- i < len ? 1 : 0;
+        (out[i] - 0) * (out[i] - 1) === 0; // out[i] must be 0 or 1
+        oneCount += out[i];
+    }
+    oneCount === len; // it can hold only when len <= N
+
+    signal from0to1[N - 1]; // from0to1[i] = (out[i] == 0) && (out[i + 1] == 1)
+    var from0to1Count = 0;
+    for (var i = 0; i < N - 1; i++) {
+        from0to1[i] <== (1 - out[i]) * out[i + 1];
+        from0to1Count += from0to1[i];
+    }
+    from0to1Count === 0; // no 0 to 1  =>  all 1 must be at the left most side
+}
+
+// Is v being included in the first prefixLen elements of in[]?
+template IncludeInPrefix(N) { // complexity 5N
+    signal input in[N];
+    signal input prefixLen; // 0 <= prefixLen <= N
+    signal input v;
+    signal output out; // 1 iff v is in  in[0 .. prefixLen - 1]
+
+    signal leadingOnes[N] <== LeadingOnes(N)(prefixLen);
+    signal isGood[N];
+    var goodCount = 0;
+    for (var i = 0; i < N; i++) {
+        isGood[i] <== AND()(IsEqual()([in[i], v]), leadingOnes[i]);
+        goodCount += isGood[i];
+    }
+    out <== IsNonZero()(goodCount);
+}
+
+// Compuates a Merkle root at rootLv made from childrens[0 .. rootLv - 1][] and leaf.
+//
+// Each childrens[i] must include the digest of childrens[i - 1] for i = 1 ... rootLv - 1
+// childrens[0] must include leaf.
+//
+// root = digest of childrens[rootLv - 1] if rootLv > 0
+// root = leaf if rootLv == 0
+template CheckMerkleProofAndComputeRoot(H, W) {
+    signal input childrens[H - 1][W]; // we only check childrens[] *below* rootLv
+    signal input rootLv; // 0 <= rootLv <= H - 1  (otherwise will be rejected)
+    signal input leaf;
+    signal output root;
+
+    // TBI: to be included
+    //
+    // childrens[lv] must include TBI[lv]  for all lv in [0 ... rootLv - 1]
+    //
+    // TBI[0] = leaf
+    // TBI[lv] = digest of children[lv - 1]  for all lv in [1 ... rootLv]
+    // TBI[rootLv] is the root
+    //
+    // For rootLv = 3, H = 5:                                                         mustInclude[]
+    //                                  TBI[4]                                        0
+    // rootLv =>                        TBI[3] <== digest of childrens[2] ==> *root*  0
+    //           childrens[2]  include  TBI[2] <== digest of childrens[1]             1
+    //           childrens[1]  include  TBI[1] <== digest of childrens[0]             1
+    //           childrens[0]  include  TBI[0] <== *leaf*                             1
+    signal TBI[H];
+    TBI[0] <== leaf;
+
+    // LeadingOnes() will also help us enforce rootLv <= H - 1
+    signal mustInclude[H - 1] <== LeadingOnes(H - 1)(rootLv);
+    signal tmp0[H - 1];
+    for (var lv = 0; lv < H - 1; lv++) {
+        // Either  childrens[lv] include TBI[lv]   or   !mustInclude[lv]
+        tmp0[lv] <== OR()(
```

Comment:
There's no need for intermediate signals here, you can directly compare against `1`.

e.g.

```
1 === OR()(Include(W)( childrens[lv], TBI[lv] ), NOT()( mustInclude[lv] ));
```